packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -0,0 +1,378 @@
+    const oneComponent = components.getSourceComponents().first();
+
+    if (components.size === 1 && this.isToolingSupported(oneComponent)) {
+      const projectNamespace = (await SfdxProjectConfig.getValue(
+        'namespace'
+      )) as string;
+      const tooling = new ToolingApi(connection, new MetadataResolver());
+      return tooling.retrieve({
+        components,
+        namespace: projectNamespace
+      });
+    }
+
+    const defaultOutput = join(
+      getRootWorkspacePath(),
+      (await SfdxPackageDirectories.getDefaultPackageDir()) ?? ''
+    );
+
+    return components
+      .retrieve({
+        usernameOrConnection: connection,
+        output: defaultOutput,
+        merge: true
+      })
+      .start();
+  }
+
+  protected async postOperation(
+    result: RetrieveResult | SourceRetrieveResult | undefined
+  ): Promise<void> {
+    if (result) {
+      const relativePackageDirs = await SfdxPackageDirectories.getPackageDirectoryPaths();
+
+      let output: string;
+
+      if (result instanceof MetadataApiRetrieveResult) {
+        output = this.createOutput(result, relativePackageDirs);
+      } else {
+        output = this.createToolingOutput(result, relativePackageDirs);
+      }
+
+      channelService.appendLine(output);
+    }
+  }
+
+  private createOutput(
+    result: MetadataApiRetrieveResult,
+    relativePackageDirs: string[]
+  ): string {
+    const successes: Row[] = [];
+    const failures: Row[] = [];
+
+    for (const response of result.getFileResponses()) {
+      const asRow = (response as unknown) as Row;
+      response.filePath = this.getRelativeProjectPath(
+        response.filePath,
+        relativePackageDirs
+      );
+      if (response.state !== ComponentStatus.Failed) {
+        successes.push(asRow);
+      } else {
+        failures.push(asRow);
+      }
+    }
+
+    return this.createOutputTable(successes, failures);
+  }
+
+  /**
+   * This exists because the Tooling API result currently doesn't conform to the
+   * same interface as the Metadata API deploy and retrieve result objects.
+   */
+  private createToolingOutput(
+    retrieveResult: SourceRetrieveResult,
+    relativePackageDirs: string[]
+  ): string {
+    const successes: Row[] = [];
+    const failures: Row[] = [];
+
+    for (const success of retrieveResult.successes) {
+      const { component, properties } = success;
+      if (component) {
+        const { fullName, type, xml } = component;
+        for (const fsPath of component.walkContent()) {
+          successes.push({
+            fullName,
+            type: type.name,
+            filePath: this.getRelativeProjectPath(fsPath, relativePackageDirs)
+          });
+        }
+        if (xml) {
+          successes.push({
+            fullName,
+            type: type.name,
+            filePath: this.getRelativeProjectPath(xml, relativePackageDirs)
+          });
+        }
+      } else if (properties) {
```

Comment:
I'm curious about testing this piece. Is this the type that would trigger this path? https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_properties.htm

Reply:
This is sort of dead code...it's a relic from a previous implementation to provide a fallback if component isn't set. I can probably get rid of it